homeassistant/components/googleactions.py

```diff
@@ -0,0 +1,390 @@
+        self.client_id = cfg.get(CONF_CLIENT_ID)
+        self.access_token = cfg.get(CONF_ACCESS_TOKEN)
+
+    @asyncio.coroutine
+    def get(self, request: Request) -> Response:
+        """Handle oauth token request."""
+        query = request.query
+        redirect_uri = query.get('redirect_uri')
+        if not redirect_uri:
+            return self.json_message(
+                "missing redirect_uri field", status_code=HTTP_BAD_REQUEST)
+
+        if self.project_id not in redirect_uri:
+            return self.json_message(
+                "missing project_id in redirect_uri",
+                status_code=HTTP_BAD_REQUEST)
+
+        state = query.get('state')
+        if not state:
+            return self.json_message(
+                "oauth request missing state", status_code=HTTP_BAD_REQUEST)
+
+        client_id = query.get('client_id')
+        if self.client_id != client_id:
+            return self.json_message(
+                "invalid client id", status_code=HTTP_UNAUTHORIZED)
+
+        generated_url = redirect_url(self.project_id, self.access_token, state)
+
+        return self.json_message(
+            "Redirect Success",
+            status_code=HTTP_MOVED_PERMANENTLY,
+            headers={'Location': generated_url})
+
+
+class GoogleActionsView(HomeAssistantView):
+    """Handle API.AI requests."""
+
+    url = GOOGLE_ACTIONS_API_ENDPOINT
+    name = 'api:googleactions'
+    requires_auth = False  # Uses access token from oauth flow
+
+    def __init__(self, hass: HomeAssistant, cfg: Dict[str, Any]) -> None:
+        """Initialize Google Actions view."""
+        super().__init__()
+
+        self.hass = hass
+        self.access_token = cfg.get(CONF_ACCESS_TOKEN)
+        self.expose_by_default = cfg.get(CONF_EXPOSE_BY_DEFAULT,
+                                         DEFAULT_EXPOSE_BY_DEFAULT)
+        self.exposed_domains = cfg.get(CONF_EXPOSED_DOMAINS,
+                                       DEFAULT_EXPOSED_DOMAINS)
+
+    def is_entity_exposed(self, entity) -> bool:
+        """Determine if an entity should be exposed on the emulated bridge.
+
+        Async friendly.
+        """
+        if entity.attributes.get('view') is not None:
+            # Ignore entities that are views
+            return False
+
+        domain = entity.domain.lower()
+        explicit_expose = entity.attributes.get(ATTR_GOOGLE_ACTIONS, None)
+
+        domain_exposed_by_default = \
+            self.expose_by_default and domain in self.exposed_domains
+
+        # Expose an entity if the entity's domain is exposed by default and
+        # the configuration doesn't explicitly exclude it from being
+        # exposed, or if the entity is explicitly exposed
+        is_default_exposed = \
+            domain_exposed_by_default and explicit_expose is not False
+
+        return is_default_exposed or explicit_expose
+
+    def handle_sync(self, hass: HomeAssistant, request_id: str):
```

Comment:
Please decouple the I/O from the handling of the messages. Have a look at the Alexa smart home skill as an example: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/alexa/smart_home.py#L32-L44

Reply:
I would even consider breaking it up in multiple files: `google_assistant/smart_home.py` and `google_assistant/http.py`